text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,210 @@
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The design is modeled on [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/) with reference to [springerle](https://github.com/carlmjohnson/springerle) -- a similar implementation in golang -- and [create-go-app](https://github.com/create-go-app/).  We explicitly do not want to re-implement cookiecutter in golang due to the scope of such an implementation.  Nor do we want to `os.Exec` a python subprocess to run cookiecutter as this would require availability of a python runtime environment.  Instead we propose to borrow heavily from create-go-app, and generate the default shell and libcnb skeletons from an embedded file system.
+
+A full invocation to generate `libcnb` scaffolding is similar to the [documented project scaffolding](https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/building-blocks-cnb/#using-the-pack-cli).
+
+```bash
+pack buildpack new --libcnb examples/ruby \
+    --api 0.7 \
+    --path ruby-buildpack \
+    --version 0.0.1 \
+    --stacks io.buildpacks.samples.stacks.bionic
+```
+
+A full session includes the terminal prompts that the project scaffolding tool asks of the end user:
+
+```bash
+pack buildpack new --libcnb examples/ruby \
+    --api 0.7 \
+    --path ruby-buildpack \
+    --version 0.0.1 \
+    --stacks io.buildpacks.samples.stacks.bionic
+Package name (often github.com/username/repo)? github.com/AidanDelaney/ruby
+Created project in ruby-buildpack
+```
+
+Templates are provided as a file tree.  A root-level `prompts.yaml` file contains prompts for the end-user.  For example, a partial file tree containing user prompts and golang skeleton code is:
+
+```bash
+.
+├── prompts.yaml
+├── cmd
+    └── main.go
+```
+
+Where `prompts.yaml` is of the form:
+
+```yaml
+prompts:
+    - variable: packageName
+      prompt: Package name (often github.com/username/repo)
+```
+
+And an example templated source file is:
+
+```golang
+package main
+
+import (
+	"github.com/buildpacks/libcnb"
+
+	bp "{{ .packageName }}/pkg"
+)
+
+func main() {
+	libcnb.Main(
+		bp.Detect{},
+		bp.Build{},
+	)
+}
+```
+
+The following template variables, known as the _reserved template variables_ are available for substitution in all project files:
+
+* `.ID` -- the buildpack id that will be written to `buildpack.toml` obtained from a positional argument to `pack`
+* `.Version` -- the buildpack version obtained from the argument to the `--version` flag
+* `.API` -- the buildpack API version obtained from the argument to the `--api` flag
+* `.Stacks` -- a slice containing the buildpack stacks obtained from the argument to the `--stacks` flag
```

Comment:
are there defaults for these values? I could imagine `version` always being `0.1.0` or `0.0.1` depending on what we want to do there.

I also imagine for simplicity a language binding may only support a single buildpack API so it in `prompt.yaml` may want to provide this.